ParseUser$State java.lang.ClassCastException: org.json.JSONObject$1 cannot be cast to java.lang.String

I updated sdk from 1.10.0 to 1.10.3, now on ParseUser.logOut() I have this error apparently at random.
happened on my development app as well as production app.
Everything was working fine before the upgrade.
if need anything more I will try to be helpful.

this is part of MyUser class

``` java

@ParseClassName("_User")
public class MyUser extends ParseUser {

//more methods

public synchronized String getUserId() {
        if (getObjectId() == null)
        try {
            save();
            } catch(ParseException e) {
            String tempUserId = preferences.getString("temp-user-id", "");
                if (tempUserId.isEmpty()) {
                    tempUserId = ("unknown-user" + randomNumber);
                    preferencesEditor.putString("temp-user-id", tempUserId).commit();
                }
                return tempUserId;
            }

        return getObjectId();
    }

public static MyUser getCurrentUser() {
        return (MyUser) ParseUser.getCurrentUser();
    }


public static void logOut() {
        unsubscribeToPush(getCurrentUser().getUserId());
        ParseUser.logOut();   //line 211
        getCurrentUser().setAppVersion();
        subscribeToPush(getCurrentUser().getUserId());
    }

private static void unsubscribeToPush(String userId) {
        ParsePush.unsubscribeInBackground("user_" + userId);
    }
}
```

```
com.mypackage E/AndroidRuntime: FATAL EXCEPTION: main
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime: java.lang.ClassCastException: org.json.JSONObject$1 cannot be cast to java.lang.String
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.ParseUser$State.sessionToken(ParseUser.java:134)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.ParseUser.logOutAsync(ParseUser.java:1000)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.ParseUser.logOutAsync(ParseUser.java:991)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController$4$1$1.then(CachedCurrentUserController.java:186)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController$4$1$1.then(CachedCurrentUserController.java:179)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task$15.run(Task.java:825)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.BoltsExecutors$ImmediateExecutor.execute(BoltsExecutors.java:105)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.completeAfterTask(Task.java:816)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.continueWithTask(Task.java:628)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.continueWithTask(Task.java:639)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task$13.then(Task.java:731)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task$13.then(Task.java:719)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task$15.run(Task.java:825)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.BoltsExecutors$ImmediateExecutor.execute(BoltsExecutors.java:105)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.completeAfterTask(Task.java:816)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.continueWithTask(Task.java:628)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.continueWithTask(Task.java:603)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.onSuccessTask(Task.java:719)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.onSuccessTask(Task.java:709)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.onSuccessTask(Task.java:743)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController$4$1.then(CachedCurrentUserController.java:179)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController$4$1.then(CachedCurrentUserController.java:176)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task$15.run(Task.java:825)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.BoltsExecutors$ImmediateExecutor.execute(BoltsExecutors.java:105)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.completeAfterTask(Task.java:816)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.continueWithTask(Task.java:628)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at bolts.Task.continueWithTask(Task.java:639)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController$4.then(CachedCurrentUserController.java:176)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController$4.then(CachedCurrentUserController.java:170)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.TaskQueue.enqueue(TaskQueue.java:69)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.CachedCurrentUserController.logOutAsync(CachedCurrentUserController.java:170)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.ParseUser.logOutInBackground(ParseUser.java:955)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.parse.ParseUser.logOut(ParseUser.java:982)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.mypackage.nova_framework.user.MyUser.logOut(MyUser.java:211)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.mypackage.info_page.InfoPageActivity$5.onClick(InfoPageActivity.java:311)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.android.internal.app.AlertController$ButtonHandler.handleMessage(AlertController.java:166)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at android.os.Handler.dispatchMessage(Handler.java:99)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at android.os.Looper.loop(Looper.java:137)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at android.app.ActivityThread.main(ActivityThread.java:4441)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at java.lang.reflect.Method.invokeNative(Native Method)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at java.lang.reflect.Method.invoke(Method.java:511)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:784)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:565)
10-21 16:33:55.585 3764-3764/com.mypackage E/AndroidRuntime:     at dalvik.system.NativeStart.main(Native Method)
```

I've seen this happen for 3 users.  The sequence of events was like this:

1. They install the app
2. The app logs an error that it can't save the ParseInstallation
3. Neither the ParseUser nor ParseInstallation get saved
4. Later the app logs an error that the ParseUser cannot be saved (JSONObject$ error above)

This is odd because my code will not save the ParseInstallation until the ParseUser is saved.  Its also odd that the ParseInstallation error comes first.

It looks to me like the ParseUser fails to save, but, saveInBackground completes successfully, letting my code move ahead to try to save the ParseInstallation, which then fails to save because I reference the ParseUser on it.

Perhaps on a subsequent app launch the app tries to save the ParseUser again, and this time an error is thrown.

Here is the ParseInstallation error:

```
Failed to save parse installation 
ex	java.lang.IllegalStateException: unable to encode an association with an unsaved ParseObject 
stacktrace	java.lang.Exception: java.lang.IllegalStateException: unable to encode an association with an unsaved ParseObject at com.aadhk.woinvoice.util.eh.b(ProGuard:165) at com.aadhk.woinvoice.util.eh.a(ProGuard:156) at a.u.run(ProGuard:825) at a.h.execute(ProGuard:105) at a.n.d(ProGuard:816) at a.n.b(ProGuard:32) at a.q.b(ProGuard:621) at a.q.a(ProGuard:618) at a.n.k(ProGuard:861) at a.n.c(ProGuard:32) at a.ab.a(ProGuard:932) at a.ab.b(ProGuard:959) at a.v.b(ProGuard:840) at a.v.a(ProGuard:829) at a.t.run(ProGuard:784) at a.h.execute(ProGuard:105) at a.n.c(ProGuard:775) at a.n.a(ProGuard:32) at a.p.b(ProGuard:567) at a.p.a(ProGuard:564) at a.n.k(ProGuard:861) at a.n.c(ProGuard:32) at a.ab.a(ProGuard:932) at a.ab.b(ProGuard:959) at a.v.b(ProGuard:840) at a.v.a(ProGuard:829) at a.t.run(ProGuard:784) at a.h.execute(ProGuard:105) at a.n.c(ProGuard:775) at a.n.a(ProGuard:32) at a.p.b(ProGuard:567) at a.p.a(ProGuard:564) at a.n.k(ProGuard:861) at a.n.c(ProGuard:32) at a.ab.a(ProGuard:932) at a.ab.b(ProGuard:959) at a.v.b(ProGuard:840) at a.v.a(ProGuard:829) at a.t.run(ProGuard:784) at a.h.execute(ProGuard:105) at a.n.c(ProGuard:775) at a.n.a(ProGuard:574) at a.n.a(ProGuard:585) at a.u.run(ProGuard:829) at a.h.execute(ProGuard:105) at a.n.d(ProGuard:816) at a.n.b(ProGuard:32) at a.q.b(ProGuard:621) at a.q.a(ProGuard:618) at a.n.k(ProGuard:861) at a.n.c(ProGuard:32) at a.ab.a(ProGuard:932) at a.ab.b(ProGuard:959) at a.v.b(ProGuard:840) at a.v.a(ProGuard:829) at a.t.run(ProGuard:784) at a.h.execute(ProGuard:105) at a.n.c(ProGuard:775) at a.n.a(ProGuard:574) at a.n.a(ProGuard:585) at a.u.run(ProGuard:829) at a.h.execute(ProGuard:105) at a.n.d(ProGuard:816) at a.n.b(ProGuard:32) at a.q.b(ProGuard:621) at a.q.a(ProGuard:618) at a.n.k(ProGuard:861) at a.n.c(ProGuard:32) at a.ab.a(ProGuard:916) at a.ab.b(ProGuard:950) at a.v.b(ProGuard:842) at a.v.a(ProGuard:829) at a.t.run(ProGuard:784) at a.h.execute(ProGuard:105) at a.n.c(ProGuard:775) at a.n.a(ProGuard:574) at a.n.a(ProGuard:585) at a.u.run(ProGuard:829) at a.h.execute(ProGuard:105) at a.n.d(ProGuard:816) at a.n.b(ProGuard:32) at a.q.b(ProGuard:621) at a.q.a(ProGuard:618) at a.n.k(ProGuard:861) at a.n.c(ProGuard:32) at a.ab.a(ProGuard:916) at a.ab.b(ProGuard:950) at a.z.b(ProGuard:477) at a.z.a(ProGuard:451) at a.t.run(ProGuard:784) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) at java.lang.Thread.run(Thread.java:818) Caused by: java.lang.IllegalStateException: unable to encode an association with an unsaved ParseObject at com.parse.rv.a(ProGuard:31) at com.parse.io.b(ProGuard:42) at com.parse.qq.b(ProGuard:27) at com.parse.io.b(ProGuard:129) at com.parse.mi.a(ProGuard:60) at com.parse.bj.a(ProGuard:61) at com.parse.kr.b(ProGuard:1640) at com.parse.kr.a(ProGuard:1635) at a.u.run(ProGuard:825) at a.h.execute(ProGuard:105) at a.n.d(ProGuard:816) at a.n.b(ProGuard:628) at a.n.b(ProGuard:639) at a.s.b(ProGuard:731) at a.s.a(ProGuard:719) at a.u.run(ProGuard:825) ... 32 more  
errid	c22143ec-05f4-4034-bc10-7856e4a89e01 
name	error 
appversion	0.1.171 
```

My code:

```
  private void initParse(final Task<ParseUser>.TaskCompletionSource onParseUser) {
    final ParseInit parseInit = new ParseInit(getApplicationContext(), accountId);
    // Unfortunately have to do this part on the UI thread
    parseInit.init();

    executor.submit(new Runnable() {
      @Override
      public void run() {
        // Initialize the parse user
        parseInit.initParseUser().onSuccessTask(new Continuation<ParseUser, Task<ParseInstallation>>() {
          @Override
          public Task<ParseInstallation> then(Task<ParseUser> task) throws Exception {
            final ParseUser parseUser = task.getResult();
            setVisitorProperty("parseUser", parseUser.getObjectId());
            onParseUser.setResult(parseUser);

            // Initialize the parse installation
            return parseInit.initParseInstallation().continueWithTask(new Continuation<ParseInstallation, Task<ParseInstallation>>() {
              @Override
              public Task<ParseInstallation> then(Task<ParseInstallation> task) throws Exception {
                if (task.isFaulted()) {
                  // Retry once, try to work around bug that causes '135 deviceType not specified' errors
                  // https://github.com/ParsePlatform/Parse-SDK-Android/issues/203#issuecomment-150045761
                  return parseInit.initParseInstallation().onSuccess(new Continuation<ParseInstallation, ParseInstallation>() {
                    @Override
                    public ParseInstallation then(Task<ParseInstallation> task) throws Exception {
                      App.logEvent(App.this, "parse-installation", "retry-success", task.getResult().getObjectId());
                      return task.getResult();
                    }
                  }, executor);
                } else {
                  return task;
                }
              }
            }, executor);
          }
        }, executor).onSuccess(new Continuation<ParseInstallation, Void>() {
          @Override
          public Void then(Task<ParseInstallation> task) throws Exception {
            final ParseInstallation parseInstallation = task.getResult();
            setVisitorProperty("parseInstall", parseInstallation.getObjectId());
            onParseInstallation(task.getResult());
            return null;
          }
        }, executor);
      }
    });
  }

/**
 * Helper to manage initializing parse
 */
public class ParseInit {
  private static final String TAG = "ParseInit";

  private final Context ctx;
  private final String accountId;

  public ParseInit(Context ctx, String accountId) { this.ctx = ctx; this.accountId = accountId; }

  /**
   * Must be called on the UI thread, otherwise app can crash on close
   */
  public void init() {
    Log.d(TAG, "init");
    final TimingLogger timings = new TimingLogger(TAG, "init");

    try {
      final boolean debug = isDebug(ctx);
      final String appId = debug ? Constant.PARSE_DEV_APP_ID : Constant.PARSE_APP_ID;
      final String clientKey = debug ? Constant.PARSE_DEV_CLIENT_KEY : Constant.PARSE_CLIENT_KEY;
      Log.d(TAG, String.format("initParse (AppId=%s, ClientKey=%s)", appId, clientKey));

      ParseObject.registerSubclass(ParseClient.class);
      ParseObject.registerSubclass(ParseItem.class);
      ParseObject.registerSubclass(ParseInvoice.class);
      ParseObject.registerSubclass(ParseAccount.class);
      ParseObject.registerSubclass(ParseGoogleAccount.class);
      ParseObject.registerSubclass(ParseMsg.class);
      ParseObject.registerSubclass(ParseSubscription.class);
      ParseObject.registerSubclass(ParseFunnel.class);
      ParseObject.registerSubclass(ParsePhoto.class);

      timings.addSplit("register");
      Parse.enableLocalDatastore(ctx);
      timings.addSplit("datastore");
      Parse.initialize(ctx, appId, clientKey);
      timings.addSplit("initialize");
    } catch (Exception e) {
      App.err(ctx, "Failed to init parse", e);
    }

    timings.dumpToLog();
  }

  public Task<ParseUser> initParseUser() {
    Log.d(TAG, "initParseUser");
    final TimingLogger timings = new TimingLogger(TAG, "initParseUser");
    ParseUser.enableAutomaticUser();
    timings.addSplit("enableAutomaticUser");

    final ParseUser user = ParseUser.getCurrentUser();
    user.put("accountId", accountId);
    try {
      user.put("countryCode", AccountUtil.getCountryCode(ctx));
    } catch (Exception e) {
      App.err(ctx, "Failed to put properties onto ParseUser", e);
    }

    return user.saveInBackground().continueWithTask(new Continuation<Void, Task<ParseUser>>() {
      @Override
      public Task<ParseUser> then(Task<Void> task) throws Exception {
        timings.addSplit("saveUser");
        if (task.isFaulted()) {
          logParseUserException(task.getError());
          return Task.forError(task.getError());
        } else {
          Log.d(TAG, "Successfully initialized parseUser: " + user.getObjectId());
          return Task.forResult(user);
        }
      }
    });
  }

  private void logParseUserException(Exception e) {
    if (e instanceof ParseException) {
      final ParseException pe = (ParseException) e;
      switch (pe.getCode()) {
        case ParseException.CONNECTION_FAILED:
          // do nothing
          break;
        default:
          // wrap it to gve us context
          App.err(ctx, "Failed to save parse user: " + pe.getCode(), new Exception(e));
          break;
      }
    } else {
      App.err(ctx, "Failed to save parse user", e);
    }
  }

  private static boolean isDebug(Context ctx) {
    // http://stackoverflow.com/questions/2718565/best-way-to-include-debug-code
    int appFlags = ctx.getApplicationInfo().flags;
    return (appFlags & ApplicationInfo.FLAG_DEBUGGABLE) != 0;
  }

  public Task<ParseInstallation> initParseInstallation() {
    final TimingLogger timings = new TimingLogger(TAG, "initParseInstallation");
    Log.d(TAG, "initParseInstallation");
    try {
      final ParseInstallation installation = ParseInstallation.getCurrentInstallation();
      installation.put("flavor", BuildConfig.FLAVOR);
      installation.put("accountId", accountId);
      installation.put("user", ParseUser.getCurrentUser());
      try {
        installation.put("androidId", Settings.Secure.getString(ctx.getContentResolver(), Settings.Secure.ANDROID_ID));
        installation.put("android", String.format("%s (%s %s)", Build.VERSION.RELEASE, Build.VERSION.CODENAME, Build.VERSION.INCREMENTAL));
        installation.put("device", String.format("%s (%s)", DeviceUtils.getDeviceName(), Build.HARDWARE));
        installation.put("abi", Build.CPU_ABI);
        installation.put("syncEnabled", Syncer.syncEnabled(ctx));
        installation.increment("appStart");
      } catch (Exception e) {
        App.err(ctx, "Failed to set properties on installation", e);
      }

      if (installation.containsKey("GCMSenderId"))
        installation.remove("GCMSenderId");

      timings.addSplit("setInstall");

      return installation.saveInBackground().continueWithTask(new Continuation<Void, Task<ParseInstallation>>() {
        @Override
        public Task<ParseInstallation> then(Task<Void> task) throws Exception {
          timings.addSplit("onInstall");
          if (task.isFaulted()) {
            if (task.getError() instanceof ParseException) {
              ParseException e = (ParseException) task.getError();
              App.err(ctx, "Failed to save parse installation: " + e.getCode(), new Exception(e));
            } else {
              App.err(ctx, "Failed to save parse installation", new Exception(task.getError()));
            }
            return Task.forError(task.getError());
          } else {
            Log.d(TAG, "Successfully initialized parseInstallation: " + installation.getObjectId());
            return Task.forResult(installation);
          }
        }
      });
    } catch (Exception e) {
      App.err(ctx, "Failed to setup parse installation", e);
      return Task.forError(e);
    }
  }
}
```

--- a/Parse/src/main/java/com/parse/ParseDecoder.java
+++ b/Parse/src/main/java/com/parse/ParseDecoder.java
@@ -71,6 +71,10 @@ import org.json.JSONObject;
       return convertJSONArrayToList((JSONArray) object);
     }
 
+    if (object == JSONObject.NULL) {
+      return null;
+    }
+
     if (!(object instanceof JSONObject)) {
       return object;
     }

--- a/Parse/src/test/java/com/parse/ParseDecoderTest.java
+++ b/Parse/src/test/java/com/parse/ParseDecoderTest.java
@@ -74,6 +74,12 @@ public class ParseDecoderTest extends ResetPluginsParseTest {
   }
 
   @Test
+  public void testNull() {
+    Object object = ParseDecoder.get().decode(JSONObject.NULL);
+    assertNull(object);
+  }
+
+  @Test
   public void testParseFieldOperations() throws JSONException {
     JSONObject json = new JSONObject();
     json.put("__op", "Increment");

